### PR TITLE
rgw/notification: trigger notifications on changes from any user

### PR DESCRIPTION
**Pull request by yuvalif**
any user authorized to make changes to a bucket may trigger
notifications defined on that bucket.
manual test procedure of the fix is described here:
https://gist.github.com/yuvalif/39c183aa0f74d286ecef7844268817df

fixes: https://tracker.ceph.com/issues/47904

Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
pubsub/multisite failures in [teuthology](http://pulpito.front.sepia.ceph.com/yuvalif-2020-11-19_12:29:38-rgw:multisite-wip-yuval-fix-47904-distro-basic-gibba/).
seems related to redundant pubsub events:
```
2020-11-19T13:15:33.989 INFO:tasks.rgw_multisite_tests:======================================================================
2020-11-19T13:15:33.989 INFO:tasks.rgw_multisite_tests:FAIL: test radosgw-admin commands
2020-11-19T13:15:33.994 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2020-11-19T13:15:33.994 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2020-11-19T13:15:33.995 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
2020-11-19T13:15:33.995 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2020-11-19T13:15:33.996 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-yuval-fix-47904/qa/tasks/rgw_multi/tests_ps.py", line 3026, in test_ps_admin
2020-11-19T13:15:33.996 INFO:tasks.rgw_multisite_tests:    assert_equal(len(parsed_result['events']), 10)
2020-11-19T13:15:33.996 INFO:tasks.rgw_multisite_tests:AssertionError: 13 != 10
```

"sync status" command that was executed before the test is showing:
```
2020-11-19T13:15:33.812 INFO:teuthology.orchestra.run.gibba038.stdout:    "full_sync": {
2020-11-19T13:15:33.813 INFO:teuthology.orchestra.run.gibba038.stdout:        "total": 1,
2020-11-19T13:15:33.813 INFO:teuthology.orchestra.run.gibba038.stdout:        "complete": 1
2020-11-19T13:15:33.813 INFO:teuthology.orchestra.run.gibba038.stdout:    }
```
could it be that we see duplicates due to full-sync? aren't we supposed to do incremental sync in the case of a pubsub zone?

**Comment by yuvalif** [comment]
rgw pubsub/multisite tests are passing in teuthology: http://pulpito.front.sepia.ceph.com/yuvalif-2020-11-25_09:13:45-rgw:multisite-wip-yuval-fix-47904-distro-basic-gibba/

File: doc/radosgw/notifications.rst [hunk 1]
```diff
@@ -50,25 +50,25 @@ In this case, the only latency added to the original operation is of committing
 Topic Management via CLI
 ------------------------
 
-Configuration of all topics of a user could be fetched using the following command:
+Configuration of all topics of (per tenant) could be fetched using the following command:
```
Comment (mattbenjamin) [inline doc/radosgw/notifications.rst:53]: associated with a tenant?

File: doc/radosgw/notifications.rst [hunk 2]
```diff
@@ -256,7 +256,7 @@ The response will have the following format:
 List Topics
 ```````````
 
-List all topics that user defined.
+List all topics (per tenant).
```
Comment (mattbenjamin) [inline doc/radosgw/notifications.rst:259]: associated with a tenant?

File: doc/radosgw/pubsub-module.rst [hunk 3]
```diff
@@ -116,52 +115,52 @@ A configuration field can be removed by using ``--tier-config-rm={key}``.
 Topic and Subscription Management via CLI
 -----------------------------------------
 
-Configuration of all topics of a user could be fetched using the following command:
+Configuration of all topics (per tenant) could be fetched using the following command:
```
Comment (mattbenjamin) [inline doc/radosgw/pubsub-module.rst:118]: ditto

File: src/rgw/rgw_pubsub.h [hunk 4]
```diff
@@ -770,7 +770,7 @@ class RGWUserPubSub
 
 
 template <class T>
-int RGWUserPubSub::read(const rgw_raw_obj& obj, T *result, RGWObjVersionTracker *objv_tracker)
+int RGWPubSub::read(const rgw_raw_obj& obj, T *result, RGWObjVersionTracker *objv_tracker)
```
Comment (mattbenjamin) [inline src/rgw/rgw_pubsub.h:773]: T* result, RGWObjVersionTracker* objv_tracker

File: src/rgw/rgw_pubsub.h [hunk 5]
```diff
@@ -793,7 +793,7 @@ int RGWUserPubSub::read(const rgw_raw_obj& obj, T *result, RGWObjVersionTracker
 }
 
 template <class T>
-int RGWUserPubSub::write(const rgw_raw_obj& obj, const T& info,
+int RGWPubSub::write(const rgw_raw_obj& obj, const T& info,
```
Comment (mattbenjamin) [inline src/rgw/rgw_pubsub.h:796]: ditto

**Review by mattbenjamin — APPROVED** [review]
all looks sane, at least to me

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
jenkins render docs

**Pull request merged**